phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/SendManager.kt

```diff
@@ -0,0 +1,619 @@
+        chain = business.chain
+    )
+
+    private val log = loggerFactory.newLogger(this::class)
+
+    sealed class BadRequestReason : Exception() {
+        object UnknownFormat : BadRequestReason()
+        object AlreadyPaidInvoice : BadRequestReason()
+        data class Expired(val timestampSeconds: Long, val expirySeconds: Long) : BadRequestReason()
+        data class ChainMismatch(val expected: Chain) : BadRequestReason()
+        data class ServiceError(val url: Url, val error: LnurlError.RemoteFailure) : BadRequestReason()
+        data class InvalidLnurl(val url: Url) : BadRequestReason()
+        data class Bip353NameNotFound(val username: String, val domain: String) : BadRequestReason()
+        data class Bip353InvalidUri(val path: String) : BadRequestReason()
+        data class Bip353InvalidOffer(val path: String) : BadRequestReason()
+        data class Bip353NoDNSSEC(val path: String) : BadRequestReason()
+        data class UnsupportedLnurl(val url: Url) : BadRequestReason()
+    }
+
+    sealed class LnurlPayError {
+        data class RemoteError(val err: LnurlError.RemoteFailure) : LnurlPayError()
+        data class BadResponseError(val err: LnurlError.Pay.Invoice) : LnurlPayError()
+        data class ChainMismatch(val expected: Chain) : LnurlPayError()
+        data object AlreadyPaidInvoice : LnurlPayError()
+    }
+
+    sealed class LnurlWithdrawError {
+        data class RemoteError(val err: LnurlError.RemoteFailure) : LnurlWithdrawError()
+    }
+
+    sealed class LnurlAuthError {
+        data class ServerError(val details: LnurlError.RemoteFailure) : LnurlAuthError()
+        data class NetworkError(val details: Throwable) : LnurlAuthError()
+        data class OtherError(val details: Throwable) : LnurlAuthError()
+    }
+
+    sealed class ParseProgress {
+        data object LnurlServiceFetch: ParseProgress()
+        data object ResolvingBip353: ParseProgress()
+    }
+
+    sealed class ParseResult {
+
+        data class BadRequest(
+            val request: String,
+            val reason: BadRequestReason
+        ): ParseResult()
+
+        data class Bolt11Invoice(
+            val request: String,
+            val invoice: fr.acinq.lightning.payment.Bolt11Invoice
+        ): ParseResult()
+
+        data class Bolt12Offer(
+            val offer: OfferTypes.Offer
+        ): ParseResult()
+
+        data class Uri(
+            val uri: BitcoinUri
+        ): ParseResult()
+
+        sealed class Lnurl: ParseResult() {
+            data class Pay(
+                val paymentIntent: LnurlPay.Intent
+            ): Lnurl()
+
+            data class Withdraw(
+                val lnurlWithdraw: LnurlWithdraw,
+            ): Lnurl()
+
+            data class Auth(
+                val auth: LnurlAuth
+            ): Lnurl()
+        }
+    }
+
+    suspend fun parse(
+        request: String,
+        progress: (p: ParseProgress) -> Unit
+    ): ParseResult {
+
```

Comment:
We should find a way to make this method a bit smarter, and fail early whenever possible.

For example: if I enter `lnbc1xxxxxx`, we should know that it's a BOLT11 invoice because it starts like one, and if the BOLT11 parsing fails, return an error immediately instead of trying everything.

Another example: if I enter something containing an `@`, it's obviously an LN address, no need to try the rest. 

And if this address does not work, we should return a nice LN address error type, so that we can display an helpful message. In many cases, we end up returning a `BadRequest(..., reason = BadRequestReason.UnknownFormat)` error, which is unhelpful and a bad UX.

I'm not sure how much work it entails. If possible, we should seize the opportunity of this PR to fix this issue. Or do you think it should be done in another PR? It would also require specific tests.

Reply:
That's a good idea. But probably best to do it in a separate PR. This was really just supposed to be a UI change, but turned into more because we dropped the MVI pattern. But at least, in terms of the code changes for this PR, the new SendManager is largely a copy-paste of pre-existing ScanController. But if we start changing a bunch of logic in the SendManager...

Reply:
Alright, let's keep things focused.